File: functions/app/domain/QuizABCD.go
package domain

import "gopkg.in/validator.v2"

// QuizABCD definition of an ABCD quiz
type QuizABCD struct {
	ID               string  `json:"id" validate:"nonzero"`
	Type         string  `json:"type" validate:"nonzero"`
	IntroHTML string  `json:"introHtml" validate:"nonzero"`
	IntroImageURL string  `json:"introImageUrl" validate:"nonzero"`
	Questions []QuizABCDQuestion  `json:"questions"`

	Answers       []uint8   `json:"answers" validate:"min=0"`
}

// Validate validates
func (quiz QuizABCD) Validate() error {
	err := validator.Validate(quiz)
	if err != nil {
		return err
	}
	for _, question := range quiz.Questions {
		err = question.Validate()
		if err != nil {
			return err
		}
	}
	return nil
}

// QuizABCDQuestion definition of an ABCD quiz question
type QuizABCDQuestion struct {
	Title         string  `json:"title" validate:"nonzero"`
	ImageURL         string  `json:"imageUrl"`
	Distractors []string  `json:"distractors"`
	CorrectNo uint8  `json:"correctNo" validate:"min=0"`
}

// Validate validates
func (question QuizABCDQuestion) Validate() error {
	return validator.Validate(question)
}

File: apps/quizabcd/app/src/domain/QuizABCD.ts
import ow from 'ow'

export interface QuizABCD {
  id: string
  type: 'abcd'
  title: string
  introHtml: string
  introImageUrl: string
  questions: QuizABCDQuestion[]
}

export interface QuizABCDQuestion {
  title: string
  imageUrl?: string
  distractors: string[]
  correctNo: number
}

export function validateQuizABCD(o: QuizABCD | any): asserts o is QuizABCD {
  ow(o, 'QuizABCD', ow.object)
  ow(o.id, 'QuizABCD.id', ow.string.nonEmpty)
  ow(o.title, 'QuizABCD.title', ow.string.nonEmpty)
  ow(o.introHtml, 'QuizABCD.introHtml', ow.string.nonEmpty)
  ow(o.introImageUrl, 'QuizABCD.introImageUrl', ow.string.nonEmpty)
  ow(o.type, 'QuizABCD.type', ow.string.equals('abcd'))
  ow(o.questions, 'QuizABCD.questions', ow.array.ofType(ow.object))
  o.questions.forEach((q: any) => validateQuizABCDQuestion(q))
}

export function validateQuizABCDQuestion(
  o: QuizABCDQuestion | any,
): asserts o is QuizABCDQuestion {
  ow(o, 'QuizABCDQuestion', ow.object)
  ow(o.title, 'QuizABCD.title', ow.string.nonEmpty)
  ow(o.imageUrl, 'QuizABCD.imageUrl', ow.any(ow.undefined, ow.string.nonEmpty))
  ow(o.distractors, 'QuizABCD.distractors', ow.array.ofType(ow.string.nonEmpty))
  ow(o.correctNo, 'QuizABCD.correctNo', ow.number.integer.finite)
}
